Added the possibility to zoom the map

Reizoukodesu opened the pull request:
Alt Key + Mouse Wheel to Zoom
All zoom levels are available in the Editor
In regular games only zooming in is allowed, the other zoom levels are
unlocked when the game is finished

Commit:
Added the possibility to zoom the map

Alt Key + Mouse Wheel to Zoom
In the Editor all zoom levels are available
In regular games only zooming in is allowed, the other zoom levels are
unlocked when the game is finished

pchote commented:
At a quick glance this looks like it deals with the issues that have caused the earlier attempted implementations to be rejected (arbitrary zoom glitches + scrolling without a modifier key).  Good job!  There are some code style issues that will need to be addressed once someone has time to give this a proper review.

Commit:
Added settings for mouse wheel zoom

The user can change the used modifier to zoom and
is able to disable the zoom feature:

Added Settings are found in the input panel:
* Allow zoom (Game.Settings.Game.AllowZoom)
* Zoom modifier (Game.Settings.Game.ZoomModifier)

pchote commented on OpenRA.Game/Graphics/Viewport.cs:
```diff
@@ -55,6 +55,12 @@ public class Viewport
+
+		public float[] AvailableZoomSteps
+		{
+			get { return new[] { 2f, 1f, 0.5f, 0.25f }; }
```
This should be a readonly field instead of a property.  Right now it creates a new object each time you call the property, which makes more work for the garbage collector.

pchote commented on OpenRA.Game/Graphics/Viewport.cs:
```diff
@@ -64,7 +70,8 @@ public float Zoom
 			set
 			{
-				zoom = value;
+				float new_value = this.ClosestTo(AvailableZoomSteps, value);
```
OpenRA code style is to omit explicit `this` unless it is required.

pchote commented on OpenRA.Game/Graphics/Viewport.cs:
```diff
@@ -74,6 +81,23 @@ public float Zoom
 
+		float ClosestTo(IEnumerable<float> collection, float target)
+		{
+			float closest_value = collection.First();
```
OpenRA code style is to use `var` instead of explicit types for local variables.

pchote commented on OpenRA.Game/Graphics/Viewport.cs:
```diff
@@ -74,6 +81,23 @@ public float Zoom
+			float closest_value = collection.First();
+			float subtract_result = Math.Abs(closest_value - target);
+
+			for (int i = 1; i < collection.Count(); i++)
```
You can use a `foreach` here instead of using `ElementAt` three times below..

pchote commented on OpenRA.Game/Graphics/Viewport.cs:
```diff
@@ -74,6 +81,23 @@ public float Zoom
 		public static int TicksSinceLastMove = 0;
 		public static int2 LastMousePos;
 
+		float ClosestTo(IEnumerable<float> collection, float target)
```
This is a private method, so there is no need to abstract away the array (which just hurts perf).  Just use `float[] collection` directly.

pchote commented on OpenRA.Mods.Common/Widgets/DropDownButtonWidget.cs:
```diff
@@ -26,6 +26,7 @@ public class DropDownButtonWidget : ButtonWidget
 		Widget panelRoot;
 
 		public string PanelRoot;
+		public object SelectedItem;
```
This shouldn't be needed here.  I'll comment later on how to avoid using this.

pchote commented on OpenRA.Mods.Common/Widgets/EditorViewportControllerWidget.cs:
```diff
@@ -76,8 +76,29 @@ public void SetTooltip(string tooltip)
 
+		void Zoom(int amount)
+		{
+			float[] zoomSteps = worldRenderer.Viewport.AvailableZoomSteps;
```
`var` instead of explicit types please.

pchote commented on OpenRA.Mods.Common/Widgets/EditorViewportControllerWidget.cs:
```diff
@@ -76,8 +76,29 @@ public void SetTooltip(string tooltip)
+			float currentZoom = worldRenderer.Viewport.Zoom;
+			int nextIndex = zoomSteps.IndexOf(currentZoom) - amount;
+			if (nextIndex < 0 || nextIndex >= zoomSteps.Count())
+			{
```
OpenRA code style also omits braces around single line blocks.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Editor/MapEditorLogic.cs:
```diff
@@ -47,9 +59,9 @@ public MapEditorLogic(Widget widget, World world, WorldRenderer worldRenderer)
+				var options = worldRenderer.Viewport.AvailableZoomSteps;
 				zoomDropdown.OnMouseDown = _ => zoomDropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 150, options, setupItem);
-				zoomDropdown.GetText = () => selectedLabel;
+				zoomDropdown.GetText = () => zoomDropdown.SelectedItem.ToString();
```
Keep this as `selectedLabel` please.  This gets invoked every render frame, so creates a lot of unnecessary objects.

pchote commented on OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:
```diff
@@ -167,8 +167,33 @@ bool IsJoystickScrolling
 
+		void Zoom(int amount)
+		{
+			float[] zoomSteps = worldRenderer.Viewport.AvailableZoomSteps;
```
`var`

pchote commented on OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:
```diff
@@ -167,8 +167,33 @@ bool IsJoystickScrolling
+			float zoom;
+			int nextIndex = zoomSteps.IndexOf(currentZoom) - amount;
+			if (nextIndex < 0 || nextIndex >= zoomSteps.Count())
+			{
```
braces

pchote commented:
The scroll sensitivity is too high for this to be useable for me - it will scroll immediately between 0.25 and 4 in the map editor, with no chance to stop in the middle.  Can you add a parameter to divide this down?

pchote commented:
Aside from that issue and the code style nits (protip: check out [Contributing](https://github.com/OpenRA/OpenRA/wiki/Contributing) and [Coding Standard](https://github.com/OpenRA/OpenRA/wiki/Coding-Standard)) this looks pretty good to me.

penev92 commented:
@Sapena are you still around?

Commit:
Merge branch 'bleed' into zoom_mouse_wheel

Commit:
Changed the selected zoom field to string

Due of performance issues.
see https://github.com/OpenRA/OpenRA/pull/9906/files#r44824249

Commit:
Fixed some coding style issues

Sapena commented:
Sorry for the late answers/changes. I hope that the coding style issues are fixed now. I commented some changes because I wasn't able to fix some mentioned issues.

Commit:
Made scroll sensitivity independent

atlimit8 commented:
`OpenRA.Mods.Common/Widgets/Logic/Editor/MapEditorLogic.cs(39,9): error CS0219: Warning as Error: The variable `selectedLabel' is assigned but its value is never used`

Commit:
Move variable definition

Commit:
Fixed map editor zoom on keypress

Commit:
Removed unused variable

atlimit8 commented:
```
OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:L206: [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.

OpenRA.Mods.Common/Widgets/ViewportControllerWidget.cs:L211: [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.

OpenRA.Mods.Common/Widgets/Logic/Editor/MapEditorLogic.cs:L53: [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.

OpenRA.Mods.Common/Widgets/Logic/Editor/MapEditorLogic.cs:L54: [SA1111] The closing parenthesis or bracket must be placed on the same line as the last parameter in the parameter list.

OpenRA.Mods.Common/Widgets/Logic/Editor/MapEditorLogic.cs:L54: [SA1009] Invalid spacing around the closing parenthesis.
```

pchote commented:
As part of a cleanup of the PR queue I am closing all prs that have had changes requested and no activity for over two weeks. You are welcome to reopen this after you are able to address the review comments above.

x-a-n-a-x commented:
@Sapena : I hope you'll find the time to finish this PR :)

pchote commented:
This may end up being superseded by proper ingame resolution controls, which are approaching the top of my todo list now that the other opengl cleanups are done.

Mailaender commented:
Superseded by #10559.
